main.go

```diff
@@ -263,31 +278,111 @@ func main() {
 				l.Fatalf("validating tripit sync command: %v", err)
 			}
 
-			go func() {
+			tripitSyncDone := make(chan struct{}, 1)
+			g.Add(func() error {
 				for {
 					if tpsync.smgr.secrets.TripitOAuthToken == "" || tpsync.smgr.secrets.TripitOAuthSecret == "" {
-						log.Print("No tripit API keys saved, not running")
-						return
-					}
-					l.Print("Running tripit sync")
-					if err := tpsync.run(ctx); err != nil {
-						l.Fatalf("error running tripit sync: %v", err)
+						metricTripitSyncErrorCount.Inc()
+						l.Print("No tripit API keys saved, not running")
+					} else {
+						l.Print("Running tripit sync")
+						if err := tpsync.run(ctx); err != nil {
+							metricTripitSyncErrorCount.Inc()
+							l.Printf("error running tripit sync: %v", err)
+						} else {
+							metricTripitSyncSuccessCount.Inc()
+						}
 					}
 
 					select {
-					case <-ctx.Done():
-						return
+					case <-tripitSyncDone:
+						return nil
 					case <-time.After(tpSyncInterval):
 						continue
 					}
 				}
-			}()
+			}, func(error) {
+				tripitSyncDone <- struct{}{}
+				log.Print("returning tripit shutdown")
+
+			})
+
+		}
 
+		mainSrv := &http.Server{
+			Addr:    listen,
+			Handler: mux,
 		}
 
-		l.Printf("Listing on %s", listen)
-		if err := http.ListenAndServe(listen, mux); err != nil {
-			l.Fatalf("Error serving: %v", err)
+		g.Add(func() error {
+			l.Printf("Listing on %s", listen)
+			if err := mainSrv.ListenAndServe(); err != nil {
+				return fmt.Errorf("serving http: %v", err)
+			}
+			return nil
+		}, func(error) {
+			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+			if err := mainSrv.Shutdown(ctx); err != nil {
+				l.Printf("shutting down main http server: %v", err)
+			}
+			log.Print("returning http shutdown")
+
+		})
+
+		if promListen != "" {
+			pm := http.NewServeMux()
+			pm.Handle("/metrics", promhttp.Handler())
+
+			metricsSrv := &http.Server{
+				Addr:    promListen,
+				Handler: pm,
+			}
+
+			g.Add(func() error {
+				l.Printf("Listing for metrics on %s", promListen)
+				if err := metricsSrv.ListenAndServe(); err != nil {
+					return fmt.Errorf("serving metrics: %v", err)
+				}
+				return nil
+			}, func(error) {
+				ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+				defer cancel()
+				if err := metricsSrv.Shutdown(ctx); err != nil {
+					log.Printf("shutting down metrics server: %v", err)
+				}
+				log.Print("returning metrics shutdown")
+
+			})
+
+			metrDone := make(chan struct{}, 1)
+			g.Add(func() error {
```

Comment:
This could be implemented as a prometheus.Collector (e.g. https://github.com/prometheus/client_golang/blob/master/prometheus/go_collector.go)